Add permission check to jump past world border (Fixes #2512)

willjallen opened the pull request:
Added permission 'essentials.jump.pastWorldBorder'
Added check if player has permission to /jump past the world border

closes #2512

willjallen committed:
Addresses issue #2512

Added permission 'essentials.jump.pastWorldBorder'
Added check if players has permission to jump past the world border

Xeyame commented:
Shouldn't all permissions be fully lowercase, and should we make the "You can not jump outside of the world border" message configurable?, Everything else looks good to me.

pop4959 commented on Essentials/src/com/earth2me/essentials/commands/Commandjump.java:
```diff
@@ -46,7 +47,12 @@ public void run(final Server server, final User user, final String commandLabel,
         charge.isAffordableFor(user);
-        user.getTeleport().teleport(loc, charge, TeleportCause.COMMAND);
+        // If the user is allowed to jump outside of the vanilla world boundary
+        if(Bukkit.getWorld(user.getWorld().getUID()).getWorldBorder().isInside(loc) || user.isAuthorized("essentials.jump.pastWorldBorder")) {
```
You should add a space between the "if" and "(" to follow the convention of the other code in this file. Also, likewise, conventionally permissions are lower case and short / to the point. Something like "essentials.jump.border" might be better.

pop4959 commented on Essentials/src/com/earth2me/essentials/commands/Commandjump.java:
```diff
@@ -46,7 +47,12 @@ public void run(final Server server, final User user, final String commandLabel,
+        // If the user is allowed to jump outside of the vanilla world boundary
+        if(Bukkit.getWorld(user.getWorld().getUID()).getWorldBorder().isInside(loc) || user.isAuthorized("essentials.jump.pastWorldBorder")) {
+            user.getTeleport().teleport(loc, charge, TeleportCause.COMMAND);
+        }else {
```
space

pop4959 commented on Essentials/src/com/earth2me/essentials/commands/Commandjump.java:
```diff
@@ -46,7 +47,12 @@ public void run(final Server server, final User user, final String commandLabel,
+        if(Bukkit.getWorld(user.getWorld().getUID()).getWorldBorder().isInside(loc) || user.isAuthorized("essentials.jump.pastWorldBorder")) {
+            user.getTeleport().teleport(loc, charge, TeleportCause.COMMAND);
+        }else {
+            user.sendMessage("You can not jump outside of the world border");
```
Messages in Essentials should be configurable. You can look at some other code or pull requests in Essentials to get an idea of how you should accomplish this. It is not too hard, and many servers will want to change the message to something else.

willjallen committed:
- Corrected formatting
- Changed essentials.jump.pastWorldBorder permission to
essentials.jump.border

- Made the jump past border user message configurable: "jumpBorder"
- Added jumpBorder default message in messages.properties

drtshock commented:
You need to add the tl keys to all the translation files. Just leave the message in English though.

mdcfe commented:
@drtshock Crowdin will add them to each file once the PR is merged and the translation is approved.

drtshock commented:
ooo

jvdoorn committed:
Added world border check to teleports.

willjallen committed:
Merge pull request #1 from Argetan/teleport-border-check

Added world border check to teleports.

Ichbinjoe commented on Essentials/src/com/earth2me/essentials/Teleport.java:
```diff
@@ -72,6 +72,18 @@ public void cooldown(boolean check) throws Exception {
     }
 
+    private void border(Location loc, IUser teleportee) throws Exception {
+        // TODO: review: should we check if the world is null? I mean I kind of assume a location always has a world.
+        if (!loc.getWorld().getWorldBorder().isInside(loc)
```
Locations are always assumed to have a valid world afaik. If they didn't we would have some other problems.

mdcfe replied:
```suggestion
        if (!loc.getWorld().getWorldBorder().isInside(loc)
```

Ichbinjoe commented on Essentials/src/com/earth2me/essentials/Teleport.java:
```diff
@@ -72,6 +72,18 @@ public void cooldown(boolean check) throws Exception {
+        // TODO: review: should we check if the world is null? I mean I kind of assume a location always has a world.
+        if (!loc.getWorld().getWorldBorder().isInside(loc)
+                && !teleportee.isAuthorized("essentials.teleport.border")) {
+            if (teleportOwner != teleportee) { // I think this will check if you are using #teleportOther()?
```
Yes! Comment needs removed.

mdcfe replied:
```suggestion
            if (teleportOwner != teleportee) {
```

Ichbinjoe reviewed (approved):
Conditional approval once comments are removed.

mdcfe reviewed (changes requested):
A couple of concerns:
* This PR will need to be updated to include AsyncTeleport once it has been merged - see #3102.
* What behaviour exactly will this change? Does this break warps or jails?

pop4959 commented:
Closing due to author inactivity. Can re-open if there are signs of life, especially addressing any concerns that have been mentioned.
